tools/skbstat.py

```diff
@@ -0,0 +1,996 @@
+		   && skb_is_ipv6(skb)) {
+		if (skb_ip6_match(skb, &filter)) {
+			return cb(ctx, args);
+		}
+	}
+
+	return 0;
+}
+
+static inline int __trace_sk_input(struct pt_regs *ctx, struct sock *sk)
+{
+	union filter_args args = { };
+
+	args.input = true;
+
+	return sk_match_and_trace(ctx, sk, &args, &record_sk);
+}
+
+static inline int __trace_skb_input(struct pt_regs *ctx, struct sk_buff *skb)
+{
+	union filter_args args = { };
+
+	args.input = true;
+
+	return skb_match_and_trace(ctx, skb, &args, &record_skb);
+}
+
+static inline int __trace_sk_output(struct pt_regs *ctx, struct sock *sk)
+{
+	union filter_args args = { };
+
+	args.output = true;
+
+	return sk_match_and_trace(ctx, sk, &args, &record_sk);
+}
+
+static inline int __trace_skb_output(struct pt_regs *ctx, struct sk_buff *skb)
+{
+	union filter_args args = { };
+
+	args.output = true;
+
+	return skb_match_and_trace(ctx, skb, &args, &record_skb);
+}
+
+int trace_sk_input(struct pt_regs *ctx, struct sock *sk)
+{
+	return __trace_sk_input(ctx, sk);
+}
+
+int trace_skb_input_arg3(struct pt_regs *ctx, void *arg1, void *arg2,
+			 struct sk_buff *skb)
+{
+	return __trace_skb_input(ctx, skb);
+}
+
+int trace_skb_input(struct pt_regs *ctx, struct sk_buff *skb)
+{
+	return __trace_skb_input(ctx, skb);
+}
+
+int trace_sk_output(struct pt_regs *ctx, struct sock *sk)
+{
+	return __trace_sk_output(ctx, sk);
+}
+
+int trace_skb_output_arg3(struct pt_regs *ctx, void *arg1, void *arg2,
+			  struct sk_buff *skb)
+{
+	return __trace_skb_output(ctx, skb);
+}
+
+int trace_skb_output(struct pt_regs *ctx, struct sk_buff *skb)
+{
+	return __trace_skb_output(ctx, skb);
+}
+
+int trace_skb_forward(struct pt_regs *ctx, struct sk_buff *skb)
+{
+	union filter_args args = { };
+
+	args.forward = true;
+
+	return skb_match_and_trace(ctx, skb, &args, &record_skb);
+}
+
+int trace_skb_drop(struct pt_regs *ctx, struct sk_buff *skb)
+{
+	union filter_args args = { };
+
+	args.drop = true;
+
+	return skb_match_and_trace(ctx, skb, &args, &record_skb_backtrace);
+}
+
+int trace_skb_consume(struct pt_regs *ctx, struct sk_buff *skb)
+{
+	union filter_args args = { };
+
+	return skb_match_and_trace(ctx, skb, &args, &record_skb_backtrace);
+}
+"""
+
+# apply filters
+bpf_text = bpf_text.replace('SADDR_FILTER', saddr_filter)
+bpf_text = bpf_text.replace('DADDR_FILTER', daddr_filter)
+bpf_text = bpf_text.replace('SPORT_FILTER', sport_filter)
+bpf_text = bpf_text.replace('DPORT_FILTER', dport_filter)
+bpf_text = bpf_text.replace('PROTO_FILTER', proto_filter)
+bpf_text = bpf_text.replace('FLAGS_FILTER', flags_filter)
+bpf_text = bpf_text.replace('IP_VERSION_FILTER', ip_version_filter)
+
+# dump ebpf and exit
+if args.ebpf:
+    print(bpf_text)
+    exit()
+
+# initialize BPF
+b = BPF(text=bpf_text)
+
+# probe list
+probe_list_input_dev = {
+    "netif_receive_skb_internal": ("trace_skb_input", 30),
+}
+
+probe_list_input_ip = {
+    "ip_error": ("trace_skb_input", 19),
```

Comment:
I see below codes that the number "19" is used for sorting. Some comments will be good to understand algorithm without digging into the implementation.

Reply:
OK, I'll add some comments here.